app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb

```diff
@@ -0,0 +1,114 @@
+  #     the +:name+,
+  #     the +:email+ of the user to add, as well as
+  #     the intended +:role+ in the course.
+  #   The provided +emails+ are NOT case sensitive.
+  # @raise [CSV::MalformedCSVError] When the file provided is invalid.
+  def parse_invitations(users)
+    result =
+      if users.is_a?(File) || users.is_a?(Tempfile)
+        parse_from_file(users)
+      else
+        parse_from_form(users)
+      end
+    result.each { |user| user[:email] = user[:email].downcase }
+  end
+
+  # Invites the users from the form submission, which reflects the actual model associations.
+  #
+  # We do not use this format in the service object because it is very clumsy.
+  #
+  # @param [Hash] users The attributes from the client.
+  # @return [Array<Hash>] Array of users to be invited
+  def parse_from_form(users)
+    users.map do |(_, value)|
+      { name: value[:name], email: value[:email], role: value[:role] }
+    end
+  end
+
+  # Loads the given file, and entries with blanks in either fields are ignored.
+  # The first row is ignored if it's a header row (contains "name, email"),
+  # else it's treated like a row of student data.
+  #
+  # This method also handles the presence of UTF-8 Byte Order Marks at the
+  # start of the file, if it exists. These are invisible characters that might
+  # be persisted as the name of the student if not caught.
+  #
+  # @param [File] file Reads the given file, in CSV format, for the name and email.
+  # @return [Array<Hash>] The array of records read from the file.
+  # @raise [CSV::MalformedCSVError] When the file provided is invalid, eg. UTF-16 encoding.
+  def parse_from_file(file)
+    [].tap do |invites|
+      CSV.foreach(file).with_index(1) do |row, row_number|
+        row[0] = remove_utf8_byte_order_mark(row[0]) if row_number == 1
+        # Ignore first row if it's a header row.
+        next if row_number == 1 && header_row?(row)
+
+        invite = parse_file_row(row)
+        invites << invite if invite
+      end
+    end
+  rescue StandardError => error
+    raise CSV::MalformedCSVError.new(error), error.message
+  end
+
+  # Returns a boolean to determine whether the row is a header row.
+  #
+  # @param[Array] row Array read from CSV file.
+  # @return [Boolean] Whether the row is a header row
+  def header_row?(row)
+    row[0].casecmp('Name') == 0 && row[1].casecmp('Email') == 0
+  end
+
+  # Parses the given CSV row (array) and returns attributes for a user invitation.
+  #   - Strip whitespaces from provided emails, if any.
+  #   - Sets the name as the given email if a name was not provided.
+  #
+  # @param [Array] row Array with 3 parameters: name, email and role respectively.
+  # @return [Hash] The parsed invitation attributes given the row.
+  def parse_file_row(row)
+    return nil if row[0].blank?
+    row.unshift(row.first) unless row.length >= 2
+
+    role = parse_file_role(row[2])
+    { name: row[0], email: row[1].strip, role: role }
```

Comment:
Are some unsuspecting users likely to do something like this? :
```
name, email, role
My Name, a@bee.co, student
```

Any harm if we also strip `name` and maybe `role`? Should we also do it for the header row to make (safely) more robust?

Reply:
👍